ledger/src/blockstore.rs

```diff
@@ -1486,13 +1618,32 @@ impl Blockstore {
             write_batch,
             shred_source,
         )?;
-        merkle_root_metas
-            .entry(erasure_set)
-            .or_insert(WorkingEntry::Dirty(MerkleRootMeta::from_shred(&shred)));
+
+        let is_new_merkle_root_meta = match merkle_root_metas.entry(erasure_set) {
+            HashMapEntry::Vacant(entry) => {
+                entry.insert(WorkingEntry::Dirty(MerkleRootMeta::from_shred(&shred)));
+                true
+            }
+            HashMapEntry::Occupied(_) => false,
+        };
```

Comment:
I feel like this whole true/false bookkeeping of if this is a new erasure-meta/merkle-root or not pretty verbose and error prone.
Might be better if we just check for consistency when these entries are inserted into blockstore. I don't know how difficult that  would be because we need to send a duplicate proof in that case.

If both erasure-meta/merkle-root structs in blockstore have a reference to the shred (type, index) which initially populated them, then  we basically need a pipeline which receives shreds, obtains erasure-meta/merkle-root from them, upserts erasure-meta/merkle-root values in blockstore and if there are inconsistencies then generates and sends duplicate proofs.

i.e. separate out the process of persisting shreds into blockstore and erasure-meta/merkle-root consistency checks.

Reply:
I think this should be possible. If we are able to add a reference to the initial shred for `ErasureMeta` (`MerkleRootMeta` already has it), then we could batch and perform all checks here:
https://github.com/anza-xyz/agave/blob/2c11b7a0f990197c9d562780c9f080b6a0754a3f/ledger/src/blockstore.rs#L1035-L1056
which still has the plumbing to send duplicate proofs.

Reply:
we could reuse the `__unused_size` field as an offset to the first received coding shred in `ErasureMeta`, and then go with this approach. 
let me know if you prefer going with this approach.

Reply:
Lets explore that idea in a separate PR and see how it looks like.
Current code has become very convoluted and risky to ship; so there might be better alternative designs.

Reply:
to be clear would you like me to continue addressing comments and merge this pr? or move this to draft and explore the alternate approach in a separate pr?

Reply:
I don't know. How do you feel like about that alternative implementation?
I am not sure which one would be simpler and less risky especially if we want to hit mainnet asap.

Reply:
okay I updated this one and will explore how the alternative will look. 
most of the helper functions will be the same, just moving the logic which performs the check to the blockstore write.